repository/src/main/java/org/alfresco/repo/security/authentication/identityservice/IdentityServiceJITProvisioningHandler.java

```diff
@@ -0,0 +1,172 @@
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import com.nimbusds.openid.connect.sdk.claims.PersonClaims;
+import com.nimbusds.openid.connect.sdk.claims.UserInfo;
+
+import org.alfresco.model.ContentModel;
+import org.alfresco.repo.security.authentication.AuthenticationUtil;
+import org.alfresco.service.cmr.security.PersonService;
+import org.alfresco.service.namespace.QName;
+import org.alfresco.service.transaction.TransactionService;
+
+/**
+ * This class handles Just in Time user provisioning. It extracts {@link OIDCUserInfo}
+ * from {@link IdentityServiceFacade.DecodedAccessToken} or {@link UserInfo}
+ * and creates a new user if it does not exist in the repository.
+ */
+public class IdentityServiceJITProvisioningHandler
+{
+    private final IdentityServiceFacade identityServiceFacade;
+    private final PersonService personService;
+    private final TransactionService transactionService;
+
+    private final Function<IdentityServiceFacade.DecodedAccessToken, Optional<? extends OIDCUserInfo>> mapTokenToUserInfoResponse = token -> {
+        Optional<String> firstName = Optional.ofNullable(token.getClaim(PersonClaims.GIVEN_NAME_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast);
+        Optional<String> lastName = Optional.ofNullable(token.getClaim(PersonClaims.FAMILY_NAME_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast);
+        Optional<String> email = Optional.ofNullable(token.getClaim(PersonClaims.EMAIL_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast);
+
+        return Optional.ofNullable(token.getClaim(PersonClaims.PREFERRED_USERNAME_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast)
+            .map(this::normalizeUserId)
+            .map(username -> new OIDCUserInfo(username, firstName.orElse(""), lastName.orElse(""), email.orElse("")));
+    };
+
+    public IdentityServiceJITProvisioningHandler(IdentityServiceFacade identityServiceFacade,
+        PersonService personService,
+        TransactionService transactionService)
+    {
+        this.identityServiceFacade = identityServiceFacade;
+        this.personService = personService;
+        this.transactionService = transactionService;
+    }
+
+    public Optional<OIDCUserInfo> extractUserInfoAndCreateUserIfNeeded(String bearerToken)
+    {
+        Optional<OIDCUserInfo> userInfoResponse = Optional.ofNullable(bearerToken)
+            .filter(Predicate.not(String::isEmpty))
+            .flatMap(token -> extractUserInfoResponseFromAccessToken(token).flatMap(
+                response -> response.username() == null || response.username().isEmpty() ?
+                    extractUserInfoResponseFromEndpoint(token) :
+                    Optional.of(response)).or(() -> extractUserInfoResponseFromEndpoint(token)));
```

Comment:
I think I understand your approach: you're first trying to extract user info from the Access Token, and if the username isn't present, you then fetch the info from the user info endpoint. This is a good strategy!

I'd suggest a small optimisation to simplify the code to make the code easier to read and maintain:
```
        Optional<OIDCUserInfo> userInfoResponse = Optional.ofNullable(bearerToken)
                    .filter(Predicate.not(String::isEmpty))
                    .flatMap(token -> extractUserInfoResponseFromAccessToken(token)
                                .filter(userInfo -> StringUtils.isNotEmpty(userInfo.username()))
                                .or(() -> extractUserInfoResponseFromEndpoint(token)));
```

Reply:
Done.